utilities/transactions/pessimistic_transaction_db.cc

```diff
@@ -600,7 +601,39 @@ void WritePreparedTxnDB::AddCommitted(uint64_t prepare_seq,
       {
         WriteLock wl(&snapshots_mutex_);
         InstrumentedMutex(db_impl_->mutex());
-        snapshots_ = db_impl_->snapshots().GetAll();
+        // We only care about snapshots lower then max
+        auto all_snapshots =
+            db_impl_->snapshots().GetAll(nullptr, max_evicted_seq);
+        auto curr_seq = db_impl_->GetLatestSequenceNumber();
+        // This is to avoid updating the snapshots_ if it already updated with a
+        // more recent vesion by a concrrent thread
+        if (curr_seq > snapshots_version_) {
+          snapshots_version_ = curr_seq;
+          // Both new and old lists are sorted and the new list is subset of the
+          // previous list plus some new items. Thus if a snapshot repeats in
+          // both new and old lists, it will appear upper in the new list. So if
+          // we simply insert the new snapshots in order, if an overwritten item
+          // is still valid in the new list is either equal to the new one or
+          // already inserted before. This guarantess that a reader will who
+          // sees the list in the middle of update, will always see all the old
+          // snaphots that are still valid.
```

Comment:
Say you have
```
old list: 10, 20, 30
```
and you have a reader iterating the list and reading snapshot 20. And you update the list to:
```
new list: 30, 40, 50
```
the reader resumes after the list is fully updated. It then reads 50 and miss 30 and 40.